internal/olm/operator/registry/index_image.go

```diff
@@ -124,14 +151,306 @@ func (c IndexImageCatalogCreator) CreateCatalog(ctx context.Context, name string
+	}
+
+	// add the upgraded bundle to the list of previous bundles so that they can be rendered with a single API call
+	c.PreviousBundles = append(c.PreviousBundles, c.BundleImage)
+	f := &FBCContext{
+		Package:     c.PackageName,
+		Refs:        c.PreviousBundles,
+		ChannelName: c.ChannelName,
+	}
+
+	// Adding the FBC "f" to the originalDeclcfg to generate a new FBC
+	declcfg, err := upgradeFBC(f, originalDeclcfg, ctx)
+	if err != nil {
+		return fmt.Errorf("error creating the upgraded FBC: %v", err)
+	}
+
+	// convert declarative config to string
+	content, err := StringifyDeclConfig(declcfg)
+
+	if err != nil {
+		return fmt.Errorf("error converting declarative config to string: %v", err)
+	}
+
+	if content == "" {
+		return errors.New("File-Based Catalog contents cannot be empty")
+	}
+
+	// validate the declarative config
+	if err = ValidateFBC(declcfg); err != nil {
+		return fmt.Errorf("error validating the generated FBC: %v", err)
+	}
+
+	log.Infof("Generated a valid Upgraded File-Based Catalog")
+
+	c.FBCContent = content
+
+	return nil
+}
+
+// upgradeFBC constructs a new File-Based Catalog from both the FBCConext object and the declarative config object. This function will check to see
+// if the FBCContext object "f" is already present in the original declarative config.
+func upgradeFBC(f *FBCContext, originalDeclCfg *declarativeconfig.DeclarativeConfig, ctx context.Context) (*declarativeconfig.DeclarativeConfig, error) {
+	var err error
+
+	// Rendering the bundle image and index image into declarative config format
+	render := action.Render{
+		Refs: f.Refs,
+	}
+
+	log.SetOutput(ioutil.Discard)
+	declcfg, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if err != nil {
+		return nil, fmt.Errorf("error in rendering the bundle and index image: %v", err)
+	}
+
+	// Ensuring a valid bundle size
+	if len(declcfg.Bundles) < 1 {
+		return nil, fmt.Errorf("bundle image should contain exactly one bundle blob")
+	}
+
+	// Checking if the existing file-based catalog (before upgrade) contains the bundle and channel that we intend to insert.
+	// If the bundle already exists, we error out. If the channel already exists, we store the index of the channel. This
+	// index will be used to access the channel from the declarative config object
+	channelExists := false
+	channelIndex := -1
+	channelHead := ""
+	for i, channel := range originalDeclCfg.Channels {
+		// Find the specific channel that the bundle needs to be inserted into
+		if channel.Name == f.ChannelName && channel.Package == f.Package {
+			channelExists = true
+			channelIndex = i
+			// Check if the CSV name is already present in the channel's entries
+			for _, entry := range channel.Entries {
+				// Our upgraded bundle image is the last element of the refs we passed in
+				if entry.Name == declcfg.Bundles[len(declcfg.Bundles)-1].Name {
+					return nil, errors.New("Bundle already exists in the Index Image")
```

Comment:
For sanity tests to pass:
```suggestion
					return nil, errors.New("bundle already exists in the Index Image")
```